File: leetcode/solutions/0993.CousinsInBinaryTree/isCousins.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// TreeNode structure for LeetCode
type TreeNode = datastructure.TreeNode

func isCousins(root *TreeNode, x int, y int) bool {
	type queueData struct {
		depth   int
		rootVal int
		node    *TreeNode
	}
	queue := []queueData{
		{1, root.Val, root.Left},
		{1, root.Val, root.Right},
	}

	type seekData struct {
		ignoreParentVal int
		depth           int
		val             int
	}
	seeking := seekData{
		0,
		root.Val,
		root.Val,
	}

	for len(queue) > 0 {
		depth, rootVal, node := queue[0].depth, queue[0].rootVal, queue[0].node
		queue = queue[1:]

		switch {
		case node == nil:
			continue
		case node.Val == seeking.val:
			return depth == seeking.depth && rootVal != seeking.ignoreParentVal
		case node.Val == x:
			seeking = seekData{
				ignoreParentVal: rootVal,
				depth:           depth,
				val:             y,
			}
		case node.Val == y:
			seeking = seekData{
				ignoreParentVal: rootVal,
				depth:           depth,
				val:             x,
			}
		default:
			depth++
			queue = append(
				queue,
				queueData{depth, node.Val, node.Left},
				queueData{depth, node.Val, node.Right},
			)
		}
	}

	return false
}

File: leetcode/solutions/0993.CousinsInBinaryTree/isCousins.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function isCousins(
  root: BinaryTreeNode<number> | null,
  x: number,
  y: number
): boolean {
  if (!root) {
    return false;
  }

  interface queueData {
    depth: number;
    parentVal: number;
    node: BinaryTreeNode<number> | null;
  }

  const queue: queueData[] = [
    { depth: 1, parentVal: root.val, node: root.left },
    { depth: 1, parentVal: root.val, node: root.right },
  ];

  interface seekData {
    depth: number;
    val: number;
    ignoreParentVal: number;
  }
  let seeking: seekData = { depth: 0, val: root.val, ignoreParentVal: 0 };

  while (queue.length > 0) {
    let { depth, parentVal, node } = queue.shift()!;

    if (!node) {
      continue;
    }
    if (node.val == seeking.val) {
      return depth === seeking.depth && parentVal !== seeking.ignoreParentVal;
    }

    if (node.val === x) {
      seeking = { depth, val: y, ignoreParentVal: parentVal };
    } else if (node.val === y) {
      seeking = { depth, val: x, ignoreParentVal: parentVal };
    } else {
      depth++;
      queue.push({ depth: depth, parentVal: node.val, node: node.left });
      queue.push({ depth: depth, parentVal: node.val, node: node.right });
    }
  }

  return false;
}
